Sanitizer error in ydb/tests/olap/ttl_tiering/data_correctness.py.TestDataCorrectness.test

ydb/tests/olap/ttl_tiering/data_correctness.py.TestDataCorrectness.test

**Build Information:**
- Version: 16ccbe8aee73119b6b3bf5af118feba55ba69799
- Build Preset: release-tsan
- Branch: prestable-25-3

**Run Command:**
```
./ya make -ttt --build "release" --sanitize="thread" -DDEBUGINFO_LINES_ONLY -F 'data_correctness.py::TestDataCorrectness::test' ydb/tests/olap/ttl_tiering
```

**Owner:** [TEAM:@ydb-platform/cs](https://github.com/orgs/ydb-platform/teams/cs)

**Test run history:** [link](https://datalens.yandex/34xnbsom67hcq?branch=prestable-25-3&full_name=ydb/tests/olap/ttl_tiering/data_correctness.py.TestDataCorrectness.test&build_type=release-tsan)

More info in [dashboard](https://datalens.yandex/4un3zdm0zcnyr)

**Logs:**
- [log](https://storage.yandexcloud.net/ydb-gh-logs/ydb-platform/ydb/Regression-run_Small_and_Medium/17708525640/ya-prestable-25-3-x86-64-tsan/try_3/artifacts/logs/ydb/tests/olap/ttl_tiering/test-results/py3test/testing_out_stuff/data_correctness.py.TestDataCorrectness.test.log)
- [logsdir](https://storage.yandexcloud.net/ydb-gh-logs/ydb-platform/ydb/Regression-run_Small_and_Medium/17708525640/ya-prestable-25-3-x86-64-tsan/try_3/test_artifacts/ydb-tests-olap-ttl_tiering.zip)

**Error Summary:**
```
teardown failed:
ydb/tests/olap/ttl_tiering/base.py:33: in teardown_class
    cls.cluster.stop()
ydb/tests/library/harness/kikimr_runner.py:639: in stop
    raise daemon.SeveralDaemonErrors(saved_exceptions)
E   ydb.tests.library.harness.daemon.SeveralDaemonErrors: Daemon failed with message: Bad exit_code..
E   Process exit_code = 100.
E   Stdout file name: 
E   /home/runner/actions_runner/_work/ydb/ydb/tmp/out/ydb/tests/olap/ttl_tiering/test-results/py3test/testing_out_stuff/data_correctness.py.TestDataCorrectness.test/cluster/node_1/stdout
E   Stderr file name: 
E   /home/runner/actions_runner/_work/ydb/ydb/tmp/out/ydb/tests/olap/ttl_tiering/test-results/py3test/testing_out_stuff/data_correctness.py.TestDataCorrectness.test/cluster/node_1/stderr
E   Stderr content:
E   
E   GRpc memory quota was set but disabled due to issues with grpc quoter, to enable it use EnableGRpcMemoryQuota option
E   Current KQP shutdown state: spent 0 seconds, not started yet
E   warning: address range table at offset 0x880 has a premature terminator entry at offset 0x890
E   ==================
E   WARNING: ThreadSanitizer: data race (pid=1286528)
E   Write of size 1 at 0x724400055861 by main thread:
E   #0 NActors::TActorSystem::Stop() /-S/ydb/library/actors/core/actorsystem.cpp:426:22 (ydbd+0x1a1cedbc) (BuildId: ea2b3091d619af82130a284fffe675e57f287956)
E   #1 NKikimr::TKikimrRunner::KikimrStop(bool) /-S/ydb/core/driver_lib/run/run.cpp:2027:22 (ydbd+0x2c1c33e1) (BuildId: ea2b3091d619af82130a284fffe675e57f287956)
E   #2 NKikimr::MainRun(NKikimr::TKikimrRunConfig const&, std::__y1::shared_ptr<NKikimr::TModuleFactories>) /-S/ydb/core/driver_lib/run/main.cpp:50:17 (ydbd+0x2a9b1b3a) (BuildId: ea2b3091d619af82130a284fffe675e57f287956)
E   #3 NKikimr::NDriverClient::TClientCommandServer::Run(NYdb::NConsoleClient::TClientCommand::TConfig&) /-S/ydb/core/driver_lib/cli_utils/cli_cmds_server.cpp:50:12 (ydbd+0x2c0f050c) (BuildId: ea2b3091d619af82130a284fffe675e57f287956)
E   #4 NYdb::NConsoleClient::TClientCommand::ValidateAndRun(NYdb::NConsoleClient::TClientCommand::TConfig&) /-S/ydb/public/lib/ydb_cli/common/command.cpp:315:16 (ydbd+0x2a9f68fa) (BuildId: ea2b3091d619af82130a284fffe675e57f287956)
E   #5 NYdb::NConsoleClient::TClientCommandTree::Run(NYdb::NConsoleClient::TClientCommand::TConfig&) /-S/ydb/public/lib/ydb_cli/common/command.cpp:512:33 (ydbd+0x2a9f930b) (BuildId: ea2b3091d619af82130a284fffe675e57f287956)
E   #6 NYdb::NConsoleClient::TClientCommand::ValidateAndRun(NYdb::NConsoleClient::TClientCommand::TConfig&) /-S/ydb/public/lib/ydb_cli/common/command.cpp:315:16 (ydbd+0x2a9f68fa) (BuildId: ea2b3091d619af82130a284fffe675e57f287956)
E   #7 NYdb::NConsoleClient::TClientCommand::Process(NYdb::NConsoleClient::TClientCommand::TConfig&) /-S/ydb/public/lib/ydb_cli/common/command.cpp:257:16 (ydbd+0x2a9f5afd) (BuildId: ea2b3091d619af82130a284fffe675e57f287956)
E   #8 NKikimr::NDriverClient::NewClient(int, char**, std::__y1::shared_ptr<NKikimr::TModuleFactories>) /-S/ydb/core/driver_lib/cli_utils/cli_cmds_root.cpp:77:26 (ydbd+0x2a9ee8c0) (BuildId: ea2b3091d619af82130a284fffe675e57f287956)
E   #9 NKikimr::Main(int, char**, std::__y1::shared_ptr<NKikimr::TModuleFactories>) /-S/ydb/core/driver_lib/run/main.cpp:151:20 (ydbd+0x2a9b5855) (BuildId: ea2b3091d619af82130a284fffe675e57f287956)
E   #10 ParameterizedMain(int, char**, std::__y1::shared_ptr<NKikimr::TModuleFactories>) /-S/ydb/core/driver_lib/run/main.cpp:201:16 (ydbd+0x2a9b6ea3) (BuildId: ea2b3091d619af82130a284fffe675e57f287956)
E   #11 main /-S/ydb/apps/ydbd/main.cpp:31:12 (ydbd+0x189dbe2e) (BuildId: ea2b3091d619af82130a284fffe675e57f287956)
E   
E   Previous read of size 1 at 0x724400055861 by thread T79:
E   #0 NActors::TActorSystem::IsStopped() /-S/ydb/library/actors/core/actorsystem.cpp:99:47 (ydbd+0x1a1cc974) (BuildId: ea2b3091d619af82130a284fffe675e57f287956)
E   #1 NKikimr::NOlap::NDataFetcher::TPortionsDataFetcher::~TPortionsDataFetcher() /-S/ydb/core/tx/columnshard/data_reader/fetcher.h:92:13 (ydbd+0x2a438bf1) (BuildId: ea2b3091d619af82130a284fffe675e57f287956)
E   #2 destroy /-S/contrib/libs/cxxsupp/libcxx/include/__memory/allocator.h:154:87 (ydbd+0x2a43410d) (BuildId: ea2b3091d619af82130a284fffe675e57f287956)
E   #3 destroy<NKikimr::NOlap::NDataFetcher::TPortionsDataFetcher, 0> /-S/contrib/libs/cxxsupp/libcxx/include/__memory/allocator_traits.h:335:9 (ydbd+0x2a43410d)
E   #4 __on_zero_shared_impl<std::__y1::allocator<NKikimr::NOlap::NDataFetcher::TPortionsDataFetcher>, 0> /-S/contrib/libs/cxxsupp/libcxx/include/__memory/shared_ptr.h:317:5 (ydbd+0x2a43410d)
E   #5 std::__y1::__shared_ptr_emplace<NKikimr::NOlap::NDataFetcher::TPortionsDataFetcher, std::__y1::allocator<NKikimr::NOlap::NDataFetcher::TPortionsDataFetcher>>::__on_zero_shared() /-S/contrib/libs/cxxsupp/libcxx/include/__memory/shared_ptr.h:320:78 (ydbd+0x2a43410d)
E   #6 __release_shared /-S/contrib/libs/cxxsupp/libcxx/include/__memory/shared_ptr.h:173:7 (ydbd+0x2a43a3d2) (BuildId: ea2b3091d619af82130a284fffe675e57f287956)
E   #7 __release_shared /-S/contrib/libs/cxxsupp/libcxx/include/__memory/shared_ptr.h:219:25 (ydbd+0x2a43a3d2)
E   #8 ~shared_ptr /-S/contrib/libs/cxxsupp/libcxx/include/__memory/shared_ptr.h:694:17 (ydbd+0x2a43a3d2)
E   #9 NKikimr::NOlap::NDataFetcher::TFetchingExecutor::~TFetchingExecutor() /-S/ydb/core/tx/columnshard/data_reader/fetching_executor.h:7:7 (ydbd+0x2a43a3d2)
E   #10 destroy /-S/contrib/libs/cxxsupp/libcxx/include/__memory/allocator.h:154:87 (ydbd+0x2a430d18) (BuildId: ea2b3091d619af82130a284fffe675e57f287956)
E   #11 destroy<NKikimr::NOlap::NDataFetcher::TFetchingExecutor, 0> /-S/contrib/libs/cxxsupp/libcxx/include/__memory/allocator_traits.h:335:9 (ydbd+0x2a430d18)
E   #12 __on_zero_shared_impl<std::__y1::allocator<NKikimr::NOlap::NDataFetcher::TFetchingExecutor>, 0> /-S/contrib/libs/cxxsupp/libcxx/include/__memory/shared_ptr.h:317:5 (ydbd+0x2a430d18)
E   #13 std::__y1::__shared_ptr_emplace<NKikimr::NOlap::NDataFetcher::TFetchingExecutor, std::__y1::allocator<NKikimr::NOlap::NDataFetcher::TFetchingExecutor>>::__on_zero_shared() /-S/contrib/libs/cxxsupp/libcxx/include/__memory/shared_ptr.h:320:78 (ydbd+0x2a430d18)
E   #14 __release_shared /-S/contrib/libs/cxxsupp/libcxx/include/__memory/shared_ptr.h:173:7 (ydbd+0x35cf11bf) (BuildId: ea2b3091d619af82130a284fffe675e57f287956)
E   #15 __release_shared /-S/contrib/libs/cxxsupp/libcxx/include/__memory/shared_ptr.h:219:25 (ydbd+0x35cf11bf)
E   #16 ~shared_ptr /-S/contrib/libs/cxxsupp/libcxx/include/__memory/shared_ptr.h:694:17 (ydbd+0x35cf11bf)
E   #17 ~TWorkerTask /-S/ydb/core/tx/conveyor_composite/service/events.h:49:7 (ydbd+0x35cf11bf)
E   #18 destroy /-S/contrib/libs/cxxsupp/libcxx/include/__memory/allocator.h:154:87 (ydbd+0x35cf11bf)
E   #19 destroy<NKikimr::NConveyorComposite::TWorkerTask, 0> /-S/contrib/libs/cxxsupp/libcxx/include/__memory/allocator_traits.h:335:9 (ydbd+0x35cf11bf)
E   #20 __base_destruct_at_end /-S/contrib/libs/cxxsupp/libcxx/include/vector:1015:7 (ydbd+0x35cf11bf)
E   #21 std::__y1::vector<NKikimr::NConveyorComposite::TWorkerTask, std::__y1::allocator<NKikimr::NConveyorComposite::TWorkerTask>>::__clear[abi:fe200000]() /-S/contrib/libs/cxxsupp/libcxx/include/vector:1009:5 (ydbd+0x35cf11bf)
E   #22 operator() /-S/contrib/libs/cxxsupp/libcxx/include/vector:547:16 (ydbd+0x35cf2662) (BuildId: ea2b3091d619af82130a284fffe675e57f287956)
E   #23 ~vector /-S/contrib/libs/cxxsupp/libcxx/include/vector:558:67 (ydbd+0x35cf2662)
E   #24 ~TEvNewTask /-S/ydb/core/tx/conveyor_composite/service/events.h:108:11 (ydbd+0x35cf2662)
E   #25 NKikimr::NConveyorComposite::TEvInternal::TEvNewTask::~TEvNewTask() /-S/ydb/core/tx/conveyor_composite/service/events.h:108:11 (ydbd+0x35cf2662)
E   #26 CheckedDelete<NActors::IEventBase> /-S/util/generic/ptr.h:36:5 (ydbd+0x1a228db6) (BuildId: ea2b3091d619af82130a284fffe675e57f287956)
E   #27 Destroy<NActors::IEventBase> /-S/util/generic/ptr.h:57:9 (ydbd+0x1a228db6)
E   #28 DoDestroy /-S/util/generic/ptr.h:376:13 (ydbd+0x1a228db6)
E   #29 ~THolder /-S/util/generic/ptr.h:306:9 ...
```

--- a/ydb/library/actors/core/actorsystem.cpp
+++ b/ydb/library/actors/core/actorsystem.cpp
@@ -372,8 +372,7 @@ namespace NActors {
 
     void TActorSystem::Start() {
         ACTORLIB_DEBUG(EDebugLevel::ActorSystem, "TActorSystem::Start");
-        Y_ABORT_UNLESS(StartExecuted == false);
-        StartExecuted = true;
+        Y_ABORT_UNLESS(!StartExecuted.exchange(true));
 
         ScheduleQueue.Reset(new NSchedulerQueue::TQueueType());
         TVector<NSchedulerQueue::TReader*> scheduleReaders;
@@ -418,12 +417,10 @@ namespace NActors {
 
     void TActorSystem::Stop() {
         ACTORLIB_DEBUG(EDebugLevel::ActorSystem, "TActorSystem::Stop");
-        if (StopExecuted || !StartExecuted) {
+        if (!StartExecuted.load() || StopExecuted.exchange(true)) {
             ACTORLIB_DEBUG(EDebugLevel::ActorSystem, "TActorSystem::Stop: already stopped");
             return;
         }
-
-        StopExecuted = true;
 
         for (auto&& fn : std::exchange(DeferredPreStop, {})) {
             fn();
@@ -439,11 +436,10 @@ namespace NActors {
     void TActorSystem::Cleanup() {
         ACTORLIB_DEBUG(EDebugLevel::ActorSystem, "TActorSystem::Cleanup");
         Stop();
-        if (CleanupExecuted || !StartExecuted) {
+        if (!StartExecuted.load() || CleanupExecuted.exchange(true)) {
             ACTORLIB_DEBUG(EDebugLevel::ActorSystem, "TActorSystem::Cleanup: already cleaned up");
             return;
         }
-        CleanupExecuted = true;
         CpuManager->Cleanup();
         Scheduler.Destroy();
         ACTORLIB_DEBUG(EDebugLevel::ActorSystem, "TActorSystem::Cleanup: cleaned up");

--- a/ydb/library/actors/core/actorsystem.h
+++ b/ydb/library/actors/core/actorsystem.h
@@ -162,9 +162,9 @@ namespace NActors {
         TMutex ProxyCreationLock;
         mutable std::vector<TActorId> DynamicProxies;
 
-        bool StartExecuted = false;
-        bool StopExecuted = false;
-        bool CleanupExecuted = false;
+        std::atomic_bool StartExecuted = false;
+        std::atomic_bool StopExecuted = false;
+        std::atomic_bool CleanupExecuted = false;
 
         std::deque<std::function<void()>> DeferredPreStop;
     public:

